c10/cuda/CUDAMallocAsyncAllocator.cpp

```diff
@@ -0,0 +1,823 @@
+  stats.reserved_bytes[static_cast<size_t>(StatType::AGGREGATE)].peak = reserved_mem_peak;
+
+  return stats;
+}
+
+void resetAccumulatedStats(int device) {
+  assertValidDevice(device);
+  TORCH_WARN("For backend:cudaMallocAsync, resetAccumulatedStats has no effect.");
+}
+
+void resetPeakStats(int device) {
+  assertValidDevice(device);
+
+  CUDAGuard g(device);
+  cudaMemPool_t mempool;
+  C10_CUDA_CHECK(cudaDeviceGetDefaultMemPool(&mempool, device));
+  // Using zero as the reset value is the method recommended by Cuda driver team.
+  // Vivek Kini says:
+  //   "Resetting to zero (which is the only valid value when setting
+  //    ReservedMemHigh) resets it to ReservedMemCurrent inside the driver
+  //   (same goes for UsedMemHigh/UsedMemCurrent)"
+  uint64_t zero = 0;
+  C10_CUDA_CHECK(cudaMemPoolSetAttribute(mempool,
+                                         cudaMemPoolAttrReservedMemHigh,
+                                         &zero));
+  C10_CUDA_CHECK(cudaMemPoolSetAttribute(mempool,
+                                         cudaMemPoolAttrUsedMemHigh,
+                                         &zero));
+}
+
+std::vector<SegmentInfo> snapshot() {
+  TORCH_CHECK(false,
+              "Calling snapshot with backend:cudaMallocAsync is not meaningful. "
+              "(For backend:native, snapshot returns a detailed summary of all "
+              "blocks tracked by the allocator, but the cudaMallocAsync backend "
+              "does not track individual blocks.)");
+  // Alternative: TORCH_WARN
+  return {};
+}
+
+// CUDAGraph interactions
+void notifyCaptureBegin(int device,
+                        CaptureId_t graph_id,
+                        MempoolId_t mempool_id) {
+  std::lock_guard<std::mutex> lk(general_mutex);
+
+  TORCH_INTERNAL_ASSERT(capture_free_streams.size() == 0);
+  TORCH_CHECK(!capture_underway,
+              "Only one capture at a time is allowed in a process.")
+  capture_underway = true;
+}
+
+void notifyCaptureAboutToEnd(int device, CaptureId_t graph_id) {
+  assertValidDevice(device);
+
+  std::lock_guard<std::mutex> lk(general_mutex);
+
+  TORCH_CHECK(capture_underway,
+              "CudaMallocAsync::notifyCaptureAboutToEnd called, "
+              "but CudaMallocAsync::capture_underway is false.");
+
+  auto capture_stream = cuda::getCurrentCUDAStream(device);
+
+  // See Note [Avoid dangling free streams during CUDA graph capture]
+  for (const auto& free_stream : capture_free_streams) {
+    // cudaEventRecord requires that the input event and stream are on the same device.
+    CUDAGuard g(free_stream.device);
+
+    // CUDACachingAllocator.cpp uses raw cuda events, as do we.
+    cudaEvent_t event;
+    C10_CUDA_CHECK(cudaEventCreateWithFlags(&event, cudaEventDisableTiming));
+    C10_CUDA_CHECK(cudaEventRecord(event, free_stream.stream));
+    C10_CUDA_CHECK(cudaStreamWaitEvent(capture_stream.stream(), event));
+    C10_CUDA_CHECK(cudaEventDestroy(event));
```

Comment:
You don't have to fix this in this PR (because I think this affects the original caching allocator as well), but now I'm kind of wondering if an error does happen in this cleanup. It seems to me like the capture data structures will be left in a wonky intermediate state. Maybe we shouldn't raise an exception if these calls fail :/